Core/src/main/java/com/plotsquared/core/queue/BlockArrayCacheScopedQueueCoordinator.java

```diff
@@ -0,0 +1,169 @@
+
+/**
+ * {@link QueueCoordinator} that caches all blocks set to it in a given array of form BlockState[][][]. An offset can be
+ * applied to blocks set to it, and the scope limited.
+ */
+public class BlockArrayCacheScopedQueueCoordinator extends ScopedQueueCoordinator {
+
+    private final BlockState[][][] blockStates;
+    private final int height;
+    private final int width;
+    private final int length;
+    private final int minY;
+    private final int maxY;
+    private final int scopeMinX;
+    private final int scopeMinZ;
+    private final int scopeMaxX;
+    private final int scopeMaxZ;
+
+    private int offsetX = 0;
+    private int offsetZ = 0;
+
+    /**
+     * Construct a new instance
+     *
+     * @param blockStates Array of form BlockState[y][x][z]. Must be fully initialised.
+     * @param minY        Minimum applicable y value. Used to account for negative world heights (e.g. -64). Inclusive
+     * @param min         Inclusive location of the minimum point to limit the scope to.
+     * @param max         Exclusive location of the maximum point to limit the scope to.
+     * @since TODO
+     */
+    public BlockArrayCacheScopedQueueCoordinator(@NonNull BlockState[][][] blockStates, int minY, Location min, Location max) {
+        super(null, min, max);
+        this.blockStates = blockStates;
+        this.height = blockStates.length;
+        this.width = blockStates[0].length;
+        this.length = blockStates[0][0].length;
+        this.minY = minY;
+        this.maxY = height + minY; // exclusive
+
+        this.scopeMinX = min.getX() & 15;
+        this.scopeMinZ = min.getZ() & 15;
+        this.scopeMaxX = scopeMinX + width;
+        this.scopeMaxZ = scopeMinZ + length;
+    }
+
+    @Override
+    public boolean setBlock(int x, final int y, int z, final @NonNull BlockState id) {
+        x += offsetX;
+        z += offsetZ;
+        if (x >= scopeMinX && x < scopeMaxX && y >= minY && y < maxY && z >= scopeMinZ && z < scopeMaxZ) {
+            blockStates[y - minY][x - scopeMinX][z - scopeMinZ] = id;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean setBlock(final int x, final int y, final int z, @NonNull final Pattern pattern) {
+        int rx = x + offsetX;
+        int rz = z + offsetZ;
+        if (rx >= scopeMinX && rx < scopeMaxX && y >= minY && y < maxY && rz >= scopeMinZ && rz < scopeMaxZ) {
+            BlockState state = pattern
+                    .applyBlock(super.getMin().getBlockVector3().add(BlockVector3.at(x, y, z)))
+                    .toImmutableState();
+            blockStates[y - minY][rx - scopeMinX][rz - scopeMinZ] = state;
+        }
+        return false;
+    }
+
+    @Override
+    public @NonNull Location getMin() {
+        return super.getMin().add(offsetX - scopeMinX, 0, offsetZ - scopeMinZ);
+    }
+
+    @Override
+    public @NonNull Location getMax() {
+        return getMin().add(15, 0, 15).withY(maxY);
```

Comment:
this looks wrong, we're not enforcing a 16x16 limit anywhere else in this class.

Reply:
This is because of the very specific use case of the class for caching generation. I do, however, think the whole thing should be basically completely rewritten (see the short essay above) which would remove the necessity for this class and go back to the ChunkQueueCoordinator/similar